### ISSUE TITLE
Per-point timestamps for LiDAR data

### ISSUE THREAD

**Issue by ajyang99** [waymo-research/waymo-open-dataset#619]
Hello, for the LiDAR data in the Perception dataset, are per-point timestamps provided? I looked through the documentation and the code but was unable to find this information, so just want to double check here. Thank you!

**Comment by VincentCa** [comment]
Hi, thanks for your question!

We provide [`timestamp_micros`](https://github.com/waymo-research/waymo-open-dataset/blob/9d0be7c4aeed2ef5e11f2e739537bf2d27cec65b/src/waymo_open_dataset/dataset.proto#L390-L393) for each frame, which gives you the time of the first LiDAR scan. Knowing the spinning direction and frequency of the top (main) LiDAR, this should allow you to get a rough time estimate for when each top LiDAR point was scanned.

You might also want to look at https://github.com/waymo-research/waymo-open-dataset/blob/9d0be7c4aeed2ef5e11f2e739537bf2d27cec65b/src/waymo_open_dataset/dataset.proto#L163-L175
In the "virtual range image" representation, the point is essentially transformed in time to a common reference point (though only accounting for ego vehicle motion, not independent motion). There are some subtleties in the ego pose alignment that you may want to consider depending on your use case.

**Comment by ajyang99** [comment]
Got it, thanks for the reply! Just to further clarify:

* Does this mean that with the rough estimate, each column in the range image will have the same timestamp, and from left to right of the image the timestamps will be uniformly distributed from beginning of the sweep up to end of the sweep?
* Wrt the virtual range image, is the virtual range image obtained by transforming each point in the raw range image with the underlying vehicle pose for that point, and then transforming to the fixed vehicle frame defined by the middle column in the range image?
* Do all LiDARs start at the same time for every sweep? If not, does this mean we can only infer LiDAR point timestamps for the top LiDAR?

Thanks!

**Comment by VincentCa** [comment]
> Does this mean that with the rough estimate, each column in the range image will have the same timestamp, and from left to right of the image the estimated timestamps will be uniformly distributed from beginning time of the sweep up to end time of the sweep?

Yes, that's correct (it won't be exact, but should result in a reasonable approximation).

> Wrt the virtual range image, is the virtual range image obtained by transforming each point in the raw range image with the underlying vehicle pose for that point, and then transforming to the fixed vehicle frame defined by the middle column in the range image?

Yes.

> Do all LiDARs start at the same time for every sweep? If not, does this mean we can only infer LiDAR point timestamps for the top LiDAR?

Unfortunately, we can not share more details on the other LiDARs. Note that we also provide the range image pose for the top LiDAR only. As mentioned in our [Waymo Open Dataset paper](https://arxiv.org/pdf/1912.04838.pdf), our LiDARs are synchronized.

Hope this helps!

**Comment by ajyang99** [comment]
This helps a lot, thanks for all the information!

**Comment by ph03** [comment]
I Just want to report that the column-based "rough" lidar point timestamp approximation leads to inaccuracies that are particularly sever at the spin boundary, with errors close to the full spin-duration (I suspect this is most likely because the original raw lidar sensor data, which doesn't correspond to a perfect column-wise measurement but exhibits a per-row mechanical azimuth offset like Hesai P128 lidars, was "coerced" into a regular range image covering the full 360deg spin).

This is an example of color-coded per point-timestamps (points are motion-compensated), going from blue->red - note the time errors right at the spin boundary:

![Image](https://github.com/user-attachments/assets/10758806-c641-496a-9213-59a203f1027b)

For most top lidar spins however, there is a more precise way to obtain more accurate per-point timestamps by inferring the per-point relative time within a spin from the linear interpolation parameter of the _available_ per-pixel poses (restricting to the translation component is sufficient), which apparently were obtained using accurate source per-point timestamps:

![Image](https://github.com/user-attachments/assets/d04aa241-6f2e-4eb3-ad4c-108264fcf2a8)

However, this is only possible for spins with sufficiently different start/end poses, as for stand-still spins no unique interpolation parameter can be extracted [in this case one can fall back to the column-wise approximation].
